### PR TITLE
Speed-up seed/dup. cleaning mkFit

### PR DESCRIPTION
**Pull request by leonardogiannini**
#### PR description:

The seed cleaning algorithm (used in initial, high pT triplet, detached quad, and triplet) and the duplicate cleaning used in the pixelLess iteration are updated to reduce the processing time, while the physics is unchanged.

- the seed cleaning is updated from an N^2 comparison of all the pairs to a comparison of the tracks that can be approximately 
compatible for removal only. The tracks are binned in phi and eta and for each track, the comparison is made only inside compatible bins

- the duplicate cleaning consisted of a nested loop comparing hits. The new algorithm uses 2 simple loops per track: 1 to map hits to tracks and the second to find all the tracks sharing hits with the one examined (similarly to [TrajectoryCleanerBySharedHits](https://github.com/cms-sw/cmssw/blob/master/TrackingTools/TrajectoryCleaning/src/TrajectoryCleanerBySharedHits.cc))

The seed cleaning update is described here:
[Jan 31 update](https://indico.cern.ch/event/1117555/contributions/4701579/attachments/2381731/4069638/mkFit%20-%20update.pdf)

The duplicate cleaning is mentioned here (p.23):
[Feb 14 update](https://indico.cern.ch/event/1128089/contributions/4735248/attachments/2390777/4086913/mkFit_TRKPOG_14February2022.pdf)

#### PR validation:

The physics is validated with TTbar PU50, where blue is before PR and red is after PR

http://legianni.web.cern.ch/legianni/plots-PR37122/

CPU profiling with igprof can be found here  

- old
[global](https://legianni.web.cern.ch/legianni/cgi-bin/igprof-navigator/mkfit-update/igreport_perf_old)
[seed cleaning](https://legianni.web.cern.ch/legianni/cgi-bin/igprof-navigator/mkfit-update/igreport_perf_old/636)
[dup cleaning](https://legianni.web.cern.ch/legianni/cgi-bin/igprof-navigator/mkfit-update/igreport_perf_old/1002)
- new
[global](https://legianni.web.cern.ch/legianni/cgi-bin/igprof-navigator/mkfit-update/igreport_perf_new)
[seed cleaning](https://legianni.web.cern.ch/legianni/cgi-bin/igprof-navigator/mkfit-update/igreport_perf_new/2139)
[dup cleaning](https://legianni.web.cern.ch/legianni/cgi-bin/igprof-navigator/mkfit-update/igreport_perf_new/2048)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cms-sw/reconstruction-l2 
it would be nice to get this in pre6

**Comment by perrotta** [comment]
urgent

> @cms-sw/reconstruction-l2 it would be nice to get this in pre6

**Comment by clacaputo** [comment]
assign @cms-sw/tracking-pog-l2

**Comment by clacaputo** [comment]
> assign @cms-sw/tracking-pog-l2

Seems like the tag didn't work.
Concerning the PR, no comments on the code, but still checking the RECO differences in the test (although they seem minimal)
About the profiling, how many events did you use @leonardogiannini ?

**Comment by leonardogiannini** [comment]
> > assign @cms-sw/tracking-pog-l2
> 
> Seems like the tag didn't work. Concerning the PR, no comments on the code, but still checking the RECO differences in the test (although they seem minimal) About the profiling, how many events did you use @leonardogiannini ?

100 evts with TTbar PU 50

**Comment by clacaputo** [comment]
assign tracking-pog-l2

**Comment by clacaputo** [comment]
+reconstruction

- mkFit seed cleaning algorithm and the duplicate cleaning updated 
- reco changes observed, but minimal
- processing time reduced

File: RecoTracker/MkFitCMS/src/MkStdSeqs.cc [hunk 1]
```diff
@@ -399,54 +429,62 @@ namespace mkfit {
...
+
       for (auto itrack = 0U; itrack < ntracks; itrack++) {
         auto &trk = tracks[itrack];
         ctheta[itrack] = 1.f / std::tan(trk.theta());
+        for (int i = 0; i < trk.nTotalHits(); ++i) {
+          if (trk.getHitIdx(i) < 0)
+            continue;
+          int a = trk.getHitLyr(i);
+          int b = trk.getHitIdx(i);
+          hitMap.insert(std::make_pair(b * 1000 + a, i > 0 ? itrack : -itrack));  //negative for first hit in trk
```
Comment (mmusich) [inline RecoTracker/MkFitCMS/src/MkStdSeqs.cc:442]: what's 1000?

**Review by mmusich — COMMENTED** [review]
not necessarily needing to be followed-up.

File: RecoTracker/MkFitCMS/src/MkStdSeqs.cc (hunk 1 above)
Comment (leonardogiannini) [reply RecoTracker/MkFitCMS/src/MkStdSeqs.cc:442]: a way to write layer and hit id in a single int.

Comment (mmusich) [reply RecoTracker/MkFitCMS/src/MkStdSeqs.cc:442]: I know, I commented already here: https://github.com/cms-sw/cmssw/pull/36546#discussion_r803464227, not sure if it's in the plans to document it.

**Comment by qliphy** [comment]
+1

**Pull request merged**